### PR TITLE
Split out a base listing view from generic IndexView

### PR DESCRIPTION
**Pull request by gasman**
An attempt to clean up the issue encountered in #10452 where IndexView is making assumptions about the nature of the items being listed (e.g. their permission checks) that don't apply to all listings.

### CODE REVIEW

**Comment by gasman** [comment]
Have now cherry-picked the remainder of #10452 with the exception of 7da8a26b5b29bace2c04485e44d8932db27db92b, to verify that this avoids the need to make the annotate_permissions step 'opt-in' on generic index views.

File: wagtail/admin/views/generic/base.py [hunk 1]
```diff
@@ -113,3 +120,199 @@ def post(self, request, *args, **kwargs):
...
+
+    def get_ordering(self):
+        ordering = self.request.GET.get("ordering", self.default_ordering)
+        if ordering not in self.get_valid_orderings():
+            ordering = self.default_ordering
+        return ordering
+
+    def get_table(self, object_list, **kwargs):
+        return self.table_class(
+            self.columns,
```
Comment (laymonage) [inline wagtail/admin/views/generic/base.py:286]: Shall we add an empty `columns = []` on the class, and/or a simple implementation of `get_columns()`? There are references to `self.columns` in this class but it has no assignment to that variable.

File: wagtail/admin/views/generic/base.py [hunk 2]
```diff
@@ -113,3 +120,199 @@ def post(self, request, *args, **kwargs):
...
+                )
+            return search_backend.search(
+                self.search_query, queryset, fields=self.search_fields
+            )
+
+        filters = {
+            field + "__icontains": self.search_query
+            for field in self.search_fields or []
+        }
+        return queryset.filter(**filters)
```
Comment (laymonage) [inline wagtail/admin/views/generic/base.py:268]: This code was extracted from `ReportView` in 4bedf6e045f6beb01eea1d00077541cffc0fc722, I didn't notice until recently but I'm pretty sure the lookups should've been chained with a `|` (`OR`) operator. Shall we change it now or do you think we should make a separate PR for it?

File: wagtail/permission_policies/base.py [hunk 3]
```diff
@@ -128,6 +128,37 @@ def users_with_any_permission_for_instance(self, actions, instance):
...
+        for instance in queryset:
+            self._update_instance_annotated_permissions(
+                instance,
+                {
+                    action: self.user_has_permission_for_instance(
+                        user, action, instance
+                    )
+                    for action in actions
+                },
+            )
```
Comment (laymonage) [inline wagtail/permission_policies/base.py:159]: Just for context as I forgot to leave a comment in my original PR:

This code is used as-is for `ModelPermissionPolicy` (which is used in snippets' `IndexView`). At first glance, it might look like this could trigger a database query for each instance. However, `ModelPermissionPolicy.user_has_permission_for_instance()` makes use of `user.has_perm()` which in turn will use Django's permission cache for the model, so this doesn't use any queries.

If we were to query the `Permission` model ourselves, it would need an extra query compared to what we're doing here.

**Review by laymonage — COMMENTED** [review]
I'm happy with the idea of extracting `BaseListingView`, thanks! I just have a few comments.

File: wagtail/admin/views/generic/base.py (hunk 1 above)
Comment (gasman) [reply wagtail/admin/views/generic/base.py:287]: Have now added a default columns definition that just displays `__str__` in https://github.com/gasman/wagtail/commit/7324cb0f2362b454e5df528589ca71acb774ff27. Was originally thinking of this as an abstract base class where subclasses would have to specify the columns in order for it to be useful (since this can be a listing of _anything_, not necessarily a model), but `__str__` works as a suitably generic default I think.

Comment (gasman) [reply wagtail/admin/views/generic/base.py:287]: Oops... looks like that breaks the logic in generic.models.IndexView, which only falls back on generating the columns from `list_display` if the class-level `columns` hasn't been set. Needs more thought...

File: wagtail/admin/views/generic/base.py (hunk 2 above)
Comment (gasman) [reply wagtail/admin/views/generic/base.py:269]: Good spot! I think that's one for a separate PR.

On reflection, I'm not sure that this logic (or indeed the rest of `search_queryset`) really belongs in BaseListingView - since this is meant to cover listings of all kinds, it's not really the right place to embed knowledge about the various mechanisms for searching models (indexed vs non-indexed) - that's more a job for generic.models.IndexView. I guess we want the search box UI to still be handled in BaseListingView, but disabled by default with `is_searchable = False`, and subclasses need to supply their own `search_queryset` method if they want to take advantage of that.

File: wagtail/admin/views/generic/base.py (hunk 1 above)
Comment (laymonage) [reply wagtail/admin/views/generic/base.py:287]: Yeah I think using `__str__` makes sense!

**Comment by gasman** [comment]
I feel like BaseListingView is still holding on to a lot of implementation baggage from generic.models.IndexView that I'd like to eliminate:

* The search logic, as per https://github.com/wagtail/wagtail/pull/10488#discussion_r1219512700
* Some awkward shifting between class-level and instance-level attributes, such as `is_searchable` and `columns`. These are not going to vary on a per-request basis (or at least, the use-cases for doing so are tenuous enough that the base class shouldn't be going to this much trouble to accommodate it) - I think this has come about because IndexView wants to be able to set defaults that depend on inspecting the model class, and an instance method was the most straightforward way to do that - it would probably be more correct to do it in a class method or a metaclass.
* Inserting the 'add' button into the header. Currently BaseListingView is using the unmodified `wagtailadmin/generic/index.html` template, which inserts a button if `add_url` and `add_item_label` are found in the template context, and generic.models.IndexView supplies these. However, considering listings more generally, the action button we put there won't necessarily be an 'add' button - for reports it's more likely to be an 'export' button, for example.
* `get_search_url` is actually returning a URL name, not an actual URL
* The attribute holding the view's own URL name (used for links-to-self such as pagination and the search form action) is called `index_url_name`, but the listing isn't necessarily going to be an index page. However, when we're working at the level of model viewsets, `index_url_name` _is_ an appropriate name, so somewhere along the line we'll need to shuffle variable names around.

**Comment by gasman** [comment]
Superseded by #10540
